src/server/console/null_console_services.cpp

```diff
@@ -0,0 +1,114 @@
+#include <boost/exception/errinfo_file_name.hpp>
+#include <boost/iostreams/device/file_descriptor.hpp>
+#include <boost/iostreams/stream.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <sstream>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+mir::NullConsoleDevice::NullConsoleDevice(VTFileOperations& fops,
+                                          std::unique_ptr<mir::Device::Observer> observer,
+                                          std::string const& dev)
+    : fops{fops}, observer{std::move(observer)}, dev{dev}
+{
+}
+
+void mir::NullConsoleDevice::on_activated()
+{
+    observer->activated(mir::Fd(fops.open(dev.c_str(), O_RDWR | O_CLOEXEC)));
+}
+
+void mir::NullConsoleDevice::on_suspended()
+{
+    observer->suspended();
+}
+
+void mir::NullConsoleServices::register_switch_handlers(graphics::EventHandlerRegister&,
+                                                        std::function<bool()> const&,
+                                                        std::function<bool()> const&)
+{
+    // do nothing since we do not switch away
+}
+
+void mir::NullConsoleServices::restore()
+{
+    // no need to restore because we were never gone
+}
+
+mir::NullConsoleServices::NullConsoleServices(std::shared_ptr<VTFileOperations> const& vtops) : fops(vtops)
+{
+}
+
+std::unique_ptr<mir::VTSwitcher> mir::NullConsoleServices::create_vt_switcher()
+{
+    BOOST_THROW_EXCEPTION((std::runtime_error{"NullConsoleServices does not support VT switching"}));
+}
+
+std::future<std::unique_ptr<mir::Device>> mir::NullConsoleServices::acquire_device(
+    int major, int minor, std::unique_ptr<mir::Device::Observer> observer)
+{
+    std::stringstream filename;
+    filename << "/sys/dev/char/" << major << ":" << minor << "/uevent";
+    mir::Fd const fd{open(filename.str().c_str(), O_RDONLY | O_CLOEXEC)};
+
+    if (fd == mir::Fd::invalid)
+    {
+        BOOST_THROW_EXCEPTION((boost::enable_error_info(std::system_error{
+                                   errno, std::system_category(), "Failed to open /sys file to discover devnode"})
+                               << boost::errinfo_file_name(filename.str())));
+    }
+
+    auto const devnode = [](auto const& fd) {
+        using namespace boost::iostreams;
+        char line_buffer[1024];
+        stream<file_descriptor_source> uevent{fd, file_descriptor_flags::never_close_handle};
+
+        while (uevent.getline(line_buffer, sizeof(line_buffer)))
+        {
+            if (strncmp(line_buffer, "DEVNAME=", strlen("DEVNAME=")) == 0)
+            {
+                return std::string{"/dev/"} + std::string{line_buffer + strlen("DEVNAME=")};
+            }
+        }
+        BOOST_THROW_EXCEPTION((std::runtime_error{"Failed to read DEVNAME"}));
+    }(fd);
+
+    std::promise<std::unique_ptr<mir::Device>> device_promise;
+    try
+    {
+        auto device = std::make_unique<mir::NullConsoleDevice>(*fops, std::move(observer), devnode);
+        device->on_activated();
```

Comment:
The other implementations also ensure that the device, if it is a DRM device, is DRM master.

I *think* it would be better to preserve that invariant here; it should be as simple as a call to `drmSetMaster` iff `major == 226` (as that's the major number of DRM devices).

Reply:
Ok. We can't actually do that. There's no way to check if you're DRM master without doing something that would fail if you weren't, like modesetting. That's not worth checking here.